Fix #20085 locale failure on certain threaded locale configurations

khwilliamson opened the pull request:
Locale initialization was failing on threaded perls on systems without the ability to query the current locale (mainly linux) and with an initialization where some categories had a different locale than others

jkeenan commented:
@bram-perl, would this p.r. address your concerns?

Leont commented:
This PR could use a more descriptive title.

khwilliamson committed:
locale.c: Remove redundant loop iteration

This loop recalculates LC_ALL on the final non-LC_ALL iteration, and
then it did an extra iteration, just to recalculate LC_ALL.  This is
unnecessary and could cause memory leaks.

bram-perl commented on locale.c:
```diff
@@ -1180,7 +1180,7 @@ S_emulate_setlocale_i(pTHX_
                  * but tis is defensive coding.  The system setlocale() returns
                  * the desired information.  Calculate LC_ALL's entry on the
                  * final iteration */
-                for (PERL_UINT_FAST8_T i = 0; i <= NOMINAL_LC_ALL_INDEX; i++) {
+                for (PERL_UINT_FAST8_T i = 0; i < NOMINAL_LC_ALL_INDEX; i++) {
```
Is the comment above still accurate?
That is:

    Calculate LC_ALL's entry on the final iteration

still correct?

khwilliamson replied:
yes

khwilliamson replied:
I will clarify the comment

bram-perl reviewed (commented):
*commit locale.c: Imnprove LC_ALL recalculation determination*:

Small typo in commit message ("Imnprove" -> "Improve")

*commit locale.c: Make sure LC_ALL gets changed when necessary *:

In the message: can you include a (direct) reference to commit  19e550d7feff0e915e0f181add94aebe4e22c38b ?

(Either in the message itself or in a foot note)

bram-perl commented on locale.c:
```diff
@@ -1014,12 +1020,10 @@ S_emulate_setlocale_i(pTHX_
-     * parameter is set to LOOPING in this circumstance, to indicate to
-     * recalculate only on the final category; 0 to indicate to not recalculate
-     * at all; and 1 to indicate to do so unconditionally */
+     * 'recalc_LC_ALL' is set to true iff this call isn't part of a tight loop
```
small typo: `iff' -> 'if'

khwilliamson replied:
iff means "if and only if"  It is standard notation in English language mathematics.  I presumed it was international.  I can expand it out, but there are more than 100 occurrences of it in our source

bram-perl replied:
in that case, just leave it as is :)

(I'm/wasn't familiar with "iff".. I don't know if I ever saw it before.. maybe I did and mentally substituted it with "if")

bram-perl commented on locale.c:
```diff
@@ -1049,9 +1049,17 @@ S_emulate_setlocale_i(pTHX_
+            * iteration in a loop.  Another iteration may have changed another
+            * category, which affects LC_ALL, so we need to continue on so as
+            * to recalculait LC_ALL */
+        && ! recalc_LC_ALL
```
What slightly worries me: part of the comment above:

     Bitter experience also
     indicates that newlocale() can free up the basis locale memory if we
     call it with the new and old being the same. */

Is there now (again) a chance of that happening?

khwilliamson replied:
I think you found a flaw.  Thank you.

khwilliamson replied:
should be fixed by latest push

khwilliamson committed:
locale.c: Change to use an enum parameter

These two functions take as input a parameter that is three-valued.
Basically, the meanings are 'true', 'false', and 'special'.  This
changes them to use an enum instead, so that the compiler will detect if
they are being called with an unanticipated value.

The parameter names now become more self-documenting.

Comments are added to better explain what 'special' means, and why it
exists.

khwilliamson committed:
locale.c: Make sure LC_ALL gets changed when necessary

This fixes #20085.

In the new POSIX API for thread-safe locales, introduced in the 2008 version
of the standard, no method was defined to find out what the current locale
is.  (That has finally been changed in the 2017 version, but I am
unaware of any boxes (accessible to me) implementing it.)

Darwin added their own extension to accomplish this, and Windows instead
used an extension to the old API to accomplish thread-safety.  And glibc
has their own, hidden extension, which it's possible to Configure Perl
to use, but, since, it's not documented, it is not enabled by default.

Perl maps the old API to the new one when compiled for threads, and
where thread-safety is available.  On Configurations where there isn't
the ability to query the current locale, it itself keeps a record of
that information, maintained from start-up of the process.

Commit 19e550d changed things so that when calling the function that
changes a category's locale, if the value to change to is the same as it
already was, it is a no-op.

But LC_ALL is precisely the sum of all the other locale categories.  If
any one of them changes, LC_ALL must be updated.  There are occasions
where all subcategories are updated in a tight loop, and, to save
unnecessary work, only the final iteration updates LC_ALL (as each
iteration would throw away the results of the previous iteration).

The blamed commit inadvertently caused the code to skip updating LC_ALL
when the final iteration wouldn't result in a change in the locale of
its category, whereas LC_ALL should have been updated on that iteration
if any previous iteration affected a change.

This new commit causes LC_ALL to always be updated on the final
iteration, regardless of what happened in the previous ones.  To only
update it when absolutely necessary, we would have to have an extra
internal state (meaning a new PL_ interpreter variable) indicating that
we are in a loop and whether changes have occurred while in that loop.
That's certainly feasible, but I don't think the extra complexity is
worth it.
